src/chain/chain.rs

```diff
@@ -683,10 +674,31 @@ impl Chain {
     }
 
     fn our_expected_elders(&self) -> BTreeMap<XorName, P2pNode> {
-        self.state
+        let mut elders: BTreeMap<_, _> = self
+            .state
             .our_joined_members()
             .map(|(name, info)| (*name, info.p2p_node.clone()))
-            .collect()
+            .collect();
+
+        // Ensure that we can still handle one node lost when relocating.
+        // Currently re-promoting relocating adult to elder is not supported.
+        // Ensure that the node we eject are the we want to relocate first.
+        let min_elders = self.elder_size() + 1;
+        let num_elders = elders.len();
+        let our_info_map = self.our_info().member_map();
+        elders.extend(
+            self.state
+                .relocate_queue
+                .iter()
+                .map(|details| details.pub_id.name())
+                .filter(|name| our_info_map.get(name).is_some())
+                .filter_map(|name| self.state.our_members.get(name))
+                .filter(|info| info.state != MemberState::Left)
+                .take(min_elders.saturating_sub(num_elders))
+                .map(|info| (*info.p2p_node.name(), info.p2p_node.clone())),
+        );
```

Comment:
I don't really get this part. What I understand is: if we have too few joined members (less than the minimum number of Elders), we add a few nodes that have been marked for relocation to the set of expected Elders. Is that correct?

If it is, I don't understand how that relates to making sure that the nodes we eject are the ones we want to relocate first. Aren't we adding the nodes that are first for relocation to the set of expected Elders? What am I missing?

Reply:
> What I understand is: if we have too few joined members (less than the minimum number of Elders), we add a few nodes that have been marked for relocation to the set of expected Elders. Is that correct?

Correct

> If it is, I don't understand how that relates to making sure that the nodes we eject are the ones we want to relocate first. Aren't we adding the nodes that are first for relocation to the set of expected Elders? What am I missing?

if we have enough Joined + Relocating nodes, that we can demote one or more of the relocating nodes, we need to choose these carefully: the demoted nodes need to be the ones that will be first relocated, i.e the one return by successive calls to `relocate_queue.pop_back`, otherwise, in `poll_relocation`, we won't be able to progress as it will be blocked on the first one it tries because it is still elder. And then we would have this Adult node that does not relocate.

Does that clarify things?

Reply:
Ah, I see - we're popping from the back of the queue, that's what I missed. Okay, that makes sense, then :+1: